File: app/src/main/java/org/schabi/newpipe/util/SponsorBlockHelper.java
package org.schabi.newpipe.util;

import android.content.Context;
import android.content.SharedPreferences;

import androidx.annotation.ColorInt;
import androidx.annotation.NonNull;
import androidx.core.content.ContextCompat;
import androidx.preference.PreferenceManager;

import org.schabi.newpipe.R;
import org.schabi.newpipe.extractor.sponsorblock.SponsorBlockCategory;
import org.schabi.newpipe.extractor.sponsorblock.SponsorBlockSegment;
import org.schabi.newpipe.extractor.stream.StreamInfo;
import org.schabi.newpipe.extractor.utils.RandomStringFromAlphabetGenerator;
import org.schabi.newpipe.views.MarkableSeekBar;
import org.schabi.newpipe.views.SeekBarMarker;

import java.security.SecureRandom;

public final class SponsorBlockHelper {
    private static final String USER_ID_ALPHABET =
            "ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789";

    private SponsorBlockHelper() {
    }

    public static String getUserId(final Context context) {
        final SharedPreferences prefs = PreferenceManager.getDefaultSharedPreferences(context);
        final String key = context.getString(R.string.sponsor_block_user_id_key);
        String userId = prefs.getString(key, null);
        if (userId == null) {
            userId = RandomStringFromAlphabetGenerator.generate(
                    USER_ID_ALPHABET, 32, new SecureRandom());
            prefs.edit().putString(key, userId).apply();
        }
        return userId;
    }

    public static void markSegments(@NonNull final Context context,
                                    @NonNull final MarkableSeekBar seekBar,
                                    @NonNull final StreamInfo streamInfo) {
        seekBar.clearMarkers();

        final SharedPreferences prefs = PreferenceManager.getDefaultSharedPreferences(context);
        if (!prefs.getBoolean(context.getString(R.string.sponsor_block_enable_key), false)) {
            return;
        }

        final SponsorBlockSegment[] sponsorBlockSegments = streamInfo.getSponsorBlockSegments();
        if (sponsorBlockSegments == null || sponsorBlockSegments.length == 0) {
            return;
        }

        final long length = streamInfo.getDuration() * 1000L;
        if (length <= 0) {
            return;
        }

        for (final SponsorBlockSegment sponsorBlockSegment : sponsorBlockSegments) {
            seekBar.seekBarMarkers.add(new SeekBarMarker(
                    sponsorBlockSegment.startTime,
                    sponsorBlockSegment.endTime,
                    length,
                    convertCategoryToColor(context, sponsorBlockSegment.category)));
        }
        seekBar.drawMarkers();
    }

    @ColorInt
    private static int convertCategoryToColor(final Context context,
                                              final SponsorBlockCategory category) {
        final int colorId = switch (category) {
            case SPONSOR -> R.color.sponsor_segment;
            case INTRO -> R.color.intro_segment;
            case OUTRO -> R.color.outro_segment;
            case INTERACTION -> R.color.interaction_segment;
            case HIGHLIGHT -> R.color.highlight_segment;
            case SELF_PROMO -> R.color.self_promo_segment;
            case NON_MUSIC -> R.color.non_music_segment;
            case PREVIEW -> R.color.preview_segment;
            case FILLER -> R.color.filler_segment;
            case PENDING -> R.color.pending_segment;
        };
        return ContextCompat.getColor(context, colorId);
    }
}

File: app/src/main/java/org/schabi/newpipe/util/SponsorBlockHelper.kt
package org.schabi.newpipe.util

import android.content.Context
import androidx.core.content.ContextCompat
import androidx.core.graphics.toColorInt
import androidx.preference.PreferenceManager
import org.schabi.newpipe.R
import org.schabi.newpipe.extractor.sponsorblock.SponsorBlockCategory
import org.schabi.newpipe.extractor.stream.StreamInfo
import org.schabi.newpipe.views.MarkableSeekBar
import org.schabi.newpipe.views.SeekBarMarker

object SponsorBlockHelper {
    @JvmStatic
    fun convertCategoryToColor(
        category: SponsorBlockCategory,
        context: Context
    ): Int {
        val key: String
        val colorStr: String?
        val prefs = PreferenceManager.getDefaultSharedPreferences(context)
        when (category) {
            SponsorBlockCategory.SPONSOR -> {
                key = context.getString(R.string.sponsor_block_category_sponsor_color_key)
                colorStr = prefs.getString(key, null)
                return colorStr?.toColorInt() ?: ContextCompat.getColor(context, R.color.sponsor_segment)
            }

            SponsorBlockCategory.INTRO -> {
                key = context.getString(R.string.sponsor_block_category_intro_color_key)
                colorStr = prefs.getString(key, null)
                return colorStr?.toColorInt() ?: ContextCompat.getColor(context, R.color.intro_segment)
            }

            SponsorBlockCategory.OUTRO -> {
                key = context.getString(R.string.sponsor_block_category_outro_color_key)
                colorStr = prefs.getString(key, null)
                return colorStr?.toColorInt() ?: ContextCompat.getColor(context, R.color.outro_segment)
            }

            SponsorBlockCategory.INTERACTION -> {
                key = context.getString(R.string.sponsor_block_category_interaction_color_key)
                colorStr = prefs.getString(key, null)
                return colorStr?.toColorInt() ?: ContextCompat.getColor(context, R.color.interaction_segment)
            }

            SponsorBlockCategory.HIGHLIGHT -> {
                key = context.getString(R.string.sponsor_block_category_highlight_color_key)
                colorStr = prefs.getString(key, null)
                return colorStr?.toColorInt() ?: ContextCompat.getColor(context, R.color.highlight_segment)
            }

            SponsorBlockCategory.SELF_PROMO -> {
                key = context.getString(R.string.sponsor_block_category_self_promo_color_key)
                colorStr = prefs.getString(key, null)
                return colorStr?.toColorInt() ?: ContextCompat.getColor(context, R.color.self_promo_segment)
            }

            SponsorBlockCategory.NON_MUSIC -> {
                key = context.getString(R.string.sponsor_block_category_non_music_color_key)
                colorStr = prefs.getString(key, null)
                return colorStr?.toColorInt() ?: ContextCompat.getColor(context, R.color.non_music_segment)
            }

            SponsorBlockCategory.PREVIEW -> {
                key = context.getString(R.string.sponsor_block_category_preview_color_key)
                colorStr = prefs.getString(key, null)
                return colorStr?.toColorInt() ?: ContextCompat.getColor(context, R.color.preview_segment)
            }

            SponsorBlockCategory.FILLER -> {
                key = context.getString(R.string.sponsor_block_category_filler_color_key)
                colorStr = prefs.getString(key, null)
                return colorStr?.toColorInt() ?: ContextCompat.getColor(context, R.color.filler_segment)
            }

            SponsorBlockCategory.PENDING -> {
                key = context.getString(R.string.sponsor_block_category_pending_color_key)
                colorStr = prefs.getString(key, null)
                return colorStr?.toColorInt() ?: ContextCompat.getColor(context, R.color.pending_segment)
            }
        }
    }

    @JvmStatic
    fun markSegments(
        context: Context,
        seekBar: MarkableSeekBar,
        streamInfo: StreamInfo
    ) {
        seekBar.clearMarkers()

        val sponsorBlockSegments = streamInfo.sponsorBlockSegments ?: return

        for (sponsorBlockSegment in sponsorBlockSegments) {
            val color = convertCategoryToColor(
                sponsorBlockSegment.category,
                context
            )

            // if null, then this category should not be marked

            // duration is in seconds, we need milliseconds
            val length = streamInfo.duration * 1000

            val seekBarMarker =
                SeekBarMarker(
                    sponsorBlockSegment.startTime,
                    sponsorBlockSegment.endTime,
                    length,
                    color
                )
            seekBar.seekBarMarkers.add(seekBarMarker)
        }

        seekBar.drawMarkers()
    }

    @JvmStatic
    fun convertCategoryToFriendlyName(
        context: Context,
        category: SponsorBlockCategory
    ): String {
        return when (category) {
            SponsorBlockCategory.SPONSOR -> context.getString(
                R.string.sponsor_block_category_sponsor
            )

            SponsorBlockCategory.INTRO -> context.getString(
                R.string.sponsor_block_category_intro
            )

            SponsorBlockCategory.OUTRO -> context.getString(
                R.string.sponsor_block_category_outro
            )

            SponsorBlockCategory.INTERACTION -> context.getString(
                R.string.sponsor_block_category_interaction
            )

            SponsorBlockCategory.HIGHLIGHT -> context.getString(
                R.string.sponsor_block_category_highlight
            )

            SponsorBlockCategory.SELF_PROMO -> context.getString(
                R.string.sponsor_block_category_self_promo
            )

            SponsorBlockCategory.NON_MUSIC -> context.getString(
                R.string.sponsor_block_category_non_music
            )

            SponsorBlockCategory.PREVIEW -> context.getString(
                R.string.sponsor_block_category_preview
            )

            SponsorBlockCategory.FILLER -> context.getString(
                R.string.sponsor_block_category_filler
            )

            SponsorBlockCategory.PENDING -> context.getString(
                R.string.sponsor_block_category_pending
            )
        }
    }

    @JvmStatic
    fun convertCategoryToSkipMessage(
        context: Context,
        category: SponsorBlockCategory
    ): String {
        return when (category) {
            SponsorBlockCategory.SPONSOR ->
                context
                    .getString(R.string.sponsor_block_skip_sponsor_toast)

            SponsorBlockCategory.INTRO ->
                context
                    .getString(R.string.sponsor_block_skip_intro_toast)

            SponsorBlockCategory.OUTRO ->
                context
                    .getString(R.string.sponsor_block_skip_outro_toast)

            SponsorBlockCategory.INTERACTION ->
                context
                    .getString(R.string.sponsor_block_skip_interaction_toast)

            SponsorBlockCategory.HIGHLIGHT -> ""

            SponsorBlockCategory.SELF_PROMO ->
                context
                    .getString(R.string.sponsor_block_skip_self_promo_toast)

            SponsorBlockCategory.NON_MUSIC ->
                context
                    .getString(R.string.sponsor_block_skip_non_music_toast)

            SponsorBlockCategory.PREVIEW ->
                context
                    .getString(R.string.sponsor_block_skip_preview_toast)

            SponsorBlockCategory.FILLER ->
                context
                    .getString(R.string.sponsor_block_skip_filler_toast)

            SponsorBlockCategory.PENDING ->
                context
                    .getString(R.string.sponsor_block_skip_pending_toast)
        }
    }
}
